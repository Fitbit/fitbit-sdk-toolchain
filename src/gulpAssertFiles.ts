import path from 'path';
import stream from 'stream';

import humanizeList from 'humanize-list';
import PluginError from 'plugin-error';
import Vinyl from 'vinyl';

const PLUGIN_NAME = 'assert-files';

export default function vinylAssertFiles(expected: string[]) {
  const expectedPaths = new Set(expected.map(path.normalize));
  return new stream.Transform({
    objectMode: true,
    transform(this: stream.Transform, file: Vinyl, _, cb) {
      if (!file.isNull()) expectedPaths.delete(file.relative);
      cb(undefined, file);
    },
    flush(this: stream.Transform, cb) {
      if (expectedPaths.size !== 0) {
        this.emit(
          'error',
          new PluginError(
            PLUGIN_NAME,
            // tslint:disable-next-line:max-line-length
            `One or more required files do not exist in the project: ${humanizeList(
              [...expectedPaths],
            )}`,
          ),
        );
      }
      cb();
    },
  });
}
