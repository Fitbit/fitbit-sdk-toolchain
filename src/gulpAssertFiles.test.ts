import { Readable } from 'stream';

import Vinyl from 'vinyl';
import gulpAssertFiles from './gulpAssertFiles';

function makeTestStream(paths: string[]) {
  const stream = new Readable({ objectMode: true });
  stream._read = () => {};

  for (const path of paths) {
    stream.push(new Vinyl({ path, contents: Buffer.alloc(0) }));
  }

  stream.push(null);
  return stream;
}

function expectAssertResult({
  expected,
  actual,
}: {
  expected: string[];
  actual?: string[];
}) {
  return expect(
    new Promise((resolve, reject) => {
      const stream = makeTestStream(actual || ['foo', 'bar']).pipe(
        gulpAssertFiles(expected),
      );
      stream.on('error', reject);
      stream.on('finish', resolve);
    }),
  );
}

it.each([
  ['one required file is missing', ['baz']],
  ['two required files are missing', ['baz', 'zab']],
  ['three required files are missing', ['baz', 'zab', 'oof']],
])('emits an error if %s', (_, expected) =>
  expectAssertResult({ expected }).rejects.toMatchSnapshot(),
);

it('resolves if all requested files are present', () =>
  expectAssertResult({
    expected: ['foo', 'bar'],
  }).resolves.toBeUndefined());
