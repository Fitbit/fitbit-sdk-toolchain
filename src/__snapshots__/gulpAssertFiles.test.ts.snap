// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emits an error if one required file is missing 1`] = `
PluginError {
  "message": "One or more required files do not exist in the project: baz",
  "name": "Error",
  "plugin": "assert-files",
}
`;

exports[`emits an error if three required files are missing 1`] = `
PluginError {
  "message": "One or more required files do not exist in the project: baz, zab and oof",
  "name": "Error",
  "plugin": "assert-files",
}
`;

exports[`emits an error if two required files are missing 1`] = `
PluginError {
  "message": "One or more required files do not exist in the project: baz and zab",
  "name": "Error",
  "plugin": "assert-files",
}
`;
